openedx/core/djangoapps/content_libraries/api.py

```diff
@@ -1051,3 +1089,295 @@ def revert_changes(library_key):
         return  # If there is no draft, no action is needed.
     LibraryBundle(library_key, ref.bundle_uuid, draft_name=DRAFT_NAME).cache.clear()
     CONTENT_LIBRARY_UPDATED.send(sender=None, library_key=library_key, update_blocks=True)
+
+
+# Import from Courseware
+# ======================
+
+
+class BaseEdxImportClient(abc.ABC):
+    """
+    Base class for all courseware import clients.
+
+    Import clients are wrappers tailored to implement the steps used in the
+    import APIs and can leverage different backends.  It is not aimed towards
+    being a generic API client for Open edX.
+    """
+
+    EXPORTABLE_BLOCK_TYPES = {
+        "drag-and-drop-v2",
+        "problem",
+        "html",
+        "video",
+    }
+
+    def __init__(self, library_key=None, library=None):
+        """
+        Initialize an import client for a library.
+
+        The method accepts either a library object or a key to a library object.
+        """
+        if bool(library_key) == bool(library):
+            raise ValueError('Provide at least one of `library_key` or '
+                             '`library`, but not both.')
+        if library is None:
+            library = ContentLibrary.objects.get_by_key(library_key)
+        self.library = library
+
+    @abc.abstractmethod
+    def get_block_data(self, block_key):
+        """
+        Get the block's OLX and static files, if any.
+        """
+
+    @abc.abstractmethod
+    def get_export_keys(self, course_key):
+        """
+        Get all exportable block keys of a given course.
+        """
+
+    @abc.abstractmethod
+    def get_block_static_data(self, asset_file):
+        """
+        Get the contents of an asset_file..
+        """
+
+    def import_block(self, modulestore_key):
+        """
+        Import a single modulestore block.
+        """
+
+        # Get or create the block in the library.
+        #
+        # To dedup blocks from different courses with the same ID, we hash the
+        # course key into the imported block.
+
+        block_data = self.get_block_data(modulestore_key)
+        course_key_hash = str(abs(hash(str(modulestore_key.course_key))))
```

Comment:
During testing today, I found an issue with this: the python `hash()` function is not a stable hash. So if you run the import command twice, it will created duplicate blocks in the library.

```
$ python -c 'print(str(abs(hash(str("course-v1:edX+DemoX+Demo_Course")))))'
7449256757677617019
$ python -c 'print(str(abs(hash(str("course-v1:edX+DemoX+Demo_Course")))))'
3804951377088993027
```

Please switch to a proper hash from `hashlib` like BLAKE2.

Second, IMHO these hashes are a bit too long. Using a real example from the edX demo course, one imported block got this ID:

```
lb:DevX:IMPORTED:problem:python_grader_3595429935507010447
```

I think a short 3-5 digit alphanumeric hash (or 7 like git uses for commit short IDs) would be more than sufficient for virtually eliminating conflicts between courses while still being human-friendly:

```
lb:DevX:IMPORTED:problem:python_grader_f8h39
```
But feel free to ignore this second part of the feedback; it's not that important.

Reply:
FYI, here's an example of the resulting usage key with the latest changes:

```
lb:OC:clean:html:b8f31cd5bb6a74c9888c478ec72a75ffc_cdwdim25pmeavdvqv
```